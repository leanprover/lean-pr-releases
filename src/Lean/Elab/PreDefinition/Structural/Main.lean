/-
Copyright (c) 2021 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura, Joachim Breitner
-/
prelude
import Lean.Elab.PreDefinition.TerminationArgument
import Lean.Elab.PreDefinition.Structural.Basic
import Lean.Elab.PreDefinition.Structural.FindRecArg
import Lean.Elab.PreDefinition.Structural.Preprocess
import Lean.Elab.PreDefinition.Structural.BRecOn
import Lean.Elab.PreDefinition.Structural.IndPred
import Lean.Elab.PreDefinition.Structural.Eqns
import Lean.Elab.PreDefinition.Structural.SmartUnfolding
import Lean.Meta.Tactic.TryThis

namespace Lean.Elab
namespace Structural
open Meta

private def getFixedPrefix (declName : Name) (xs : Array Expr) (value : Expr) : MetaM Nat := do
  let numFixedRef ← IO.mkRef xs.size
  forEachExpr' value fun e => do
    if e.isAppOf declName then
      let args := e.getAppArgs
      numFixedRef.modify fun numFixed => if args.size < numFixed then args.size else numFixed
      for arg in args, x in xs do
        /- We should not use structural equality here. For example, given the definition
           ```
           def V.map {α β} f x x_1 :=
             @V.map.match_1.{1} α (fun x x_2 => V β x) x x_1
               (fun x x_2 => @V.mk₁ β x (f Bool.true x_2))
               (fun e => @V.mk₂ β (V.map (fun b => α b) (fun b => β b) f Bool.false e))
           ```
           The first three arguments at `V.map (fun b => α b) (fun b => β b) f Bool.false e` are "fixed"
           modulo definitional equality.

           We disable to proof irrelevance to be able to use structural recursion on inductive predicates.
           For example, consider the example
           ```
           inductive PList (α : Type) : Prop
           | nil
           | cons : α → PList α → PList α

           infixr:67 " ::: " => PList.cons

           set_option trace.Elab.definition.structural true in
           def pmap {α β} (f : α → β) : PList α → PList β
             | PList.nil => PList.nil
             | a:::as => f a ::: pmap f as
           ```
          The "Fixed" prefix would be 4 since all elements of type `PList α` are definitionally equal.
        -/
        if !(← withoutProofIrrelevance <| withReducible <| isDefEq arg x) then
          -- We continue searching if e's arguments are not a prefix of `xs`
          return true
      return false
    else
      return true
  numFixedRef.get

partial def withCommonTelescope (preDefs : Array PreDefinition) (k : Array Expr → Array Expr → M α) : M α :=
  go #[] (preDefs.map (·.value))
where
  go (fvars : Array Expr) (vals : Array Expr) : M α := do
    if !(vals.all fun val => val.isLambda) then
      k fvars vals
    else if !(← vals.allM fun val => return val.bindingName! == vals[0]!.bindingName! && val.binderInfo == vals[0]!.binderInfo && (← isDefEq val.bindingDomain! vals[0]!.bindingDomain!)) then
      k fvars vals
    else
      withLocalDecl vals[0]!.bindingName! vals[0]!.binderInfo vals[0]!.bindingDomain! fun x =>
        go (fvars.push x) (vals.map fun val => val.bindingBody!.instantiate1 x)

def getMutualFixedPrefix (preDefs : Array PreDefinition) : M Nat :=
  withCommonTelescope preDefs fun xs vals => do
    let resultRef ← IO.mkRef xs.size
    for val in vals do
      if (← resultRef.get) == 0 then return 0
      forEachExpr' val fun e => do
        if preDefs.any fun preDef => e.isAppOf preDef.declName then
          let args := e.getAppArgs
          resultRef.modify (min args.size ·)
          for arg in args, x in xs do
            if !(← withoutProofIrrelevance <| withReducible <| isDefEq arg x) then
              -- We continue searching if e's arguments are not a prefix of `xs`
              return true
          return false
        else
          return true
    resultRef.get

/-- Checks that all parameter types are mutually inductive -/
private def checkAllFromSameClique (recArgInfos : Array RecArgInfo) : MetaM Unit := do
  for recArgInfo in recArgInfos do
    unless recArgInfos[0]!.indAll.contains recArgInfo.indName do
      throwError m!"Cannot use structural mutual recursion: The recursive argument of " ++
        m!"{recArgInfos[0]!.fnName} is of type {recArgInfos[0]!.indName}, " ++
        m!"the recursive argument of {recArgInfo.fnName} is of type " ++
            m!"{recArgInfo.indName}, and these are not mutually recursive."

private def elimMutualRecursion (preDefs : Array PreDefinition) (recArgPoss : Array Nat) : M (Array PreDefinition) := do
  withoutModifyingEnv do
    preDefs.forM (addAsAxiom ·)
    let names := preDefs.map (·.declName)
    let preDefs ← preDefs.mapM fun preDef =>
      return { preDef with value := (← preprocess preDef.value names) }

    -- The syntactically fixed arguments
    let maxNumFixed ← getMutualFixedPrefix preDefs

    -- We do two passes to get the RecArgInfo values.
    -- From the first pass, we only keep the mininum of the `numFixed` reported.
    let numFixed ← lambdaBoundedTelescope preDefs[0]!.value maxNumFixed fun xs _ => do
      assert! xs.size = maxNumFixed
      let values ← preDefs.mapM (instantiateLambda ·.value xs)

      let recArgInfos ← preDefs.mapIdxM fun i preDef => do
        let recArgPos := recArgPoss[i]!
        let value := values[i]!
        lambdaTelescope value fun ys _value => do
          getRecArgInfo preDef.declName maxNumFixed (xs ++ ys) recArgPos

      -- Check that the inductive parameters agree. This also ensures that they only depend on the
      -- trimmed prefix (minimum of all `.numFixed`).
      for recArgInfo in recArgInfos[1:] do
        let ok ← Array.allM
          (fun (e₁, e₂) => isDefEqGuarded e₁ e₂)
          (Array.zip recArgInfo.indParams recArgInfos[0]!.indParams)
        unless recArgInfos[0]!.indParams.size = recArgInfo.indParams.size && ok do
          throwError m!"The inductive type of the recursive parameter of {recArgInfos[0]!.fnName} " ++
            m!"and {recArgInfo.fnName} have different parameters:" ++
            indentD m!"{recArgInfos[0]!.indParams}" ++
            indentD m!"{recArgInfo.indParams}"

      return (recArgInfos.map (·.numFixed)).foldl Nat.min maxNumFixed

    if numFixed < maxNumFixed then
      trace[Elab.definition.structural] "Reduced numFixed from {maxNumFixed} to {numFixed}"

    -- Now we bring exactly that `numFixed` parameter into scope.
    lambdaBoundedTelescope preDefs[0]!.value numFixed fun xs _ => do
      assert! xs.size = numFixed
      let values ← preDefs.mapM (instantiateLambda ·.value xs)

      let recArgInfos ← preDefs.mapIdxM fun i preDef => do
        let recArgPos := recArgPoss[i]!
        let value := values[i]!
        lambdaTelescope value fun ys _value => do
          getRecArgInfo preDef.declName numFixed (xs ++ ys) recArgPos

      -- Two passes should suffice
      assert! recArgInfos.all (·.numFixed = numFixed)

      let indInfo ← getConstInfoInduct recArgInfos[0]!.indName
      if ← isInductivePredicate indInfo.name then
        -- Here we branch off to the IndPred construction, but only for non-mutual functions
        unless preDefs.size = 1 do
          throwError "structural mutual recursion over inductive predicates is not supported"
        trace[Elab.definition.structural] "Using mkIndPred construction"
        let preDef := preDefs[0]!
        let recArgInfo := recArgInfos[0]!
        let value := values[0]!
        let valueNew ← mkIndPredBRecOn recArgInfo value
        let valueNew ← mkLambdaFVars xs valueNew
        trace[Elab.definition.structural] "Nonrecursive value:{indentExpr valueNew}"
        check valueNew
        return #[{ preDef with value := valueNew }]

      checkAllFromSameClique recArgInfos
      -- Sort the (indices of the) definitions by their position in indInfo.all
      let positions : Positions := .groupAndSort (·.indName) recArgInfos indInfo.all.toArray

      -- Construct the common `.brecOn` arguments
      let motives ← (Array.zip recArgInfos values).mapM fun (r, v) => mkBRecOnMotive r v
      let brecOnConst ← mkBRecOnConst recArgInfos positions motives
      let FTypes ← inferBRecOnFTypes recArgInfos positions brecOnConst
      let FArgs ← (recArgInfos.zip  (values.zip FTypes)).mapM fun (r, (v, t)) =>
        mkBRecOnF recArgInfos positions r v t
      -- Assemble the individual `.brecOn` applications
      let valuesNew ← (Array.zip recArgInfos values).mapIdxM fun i (r, v) =>
        mkBrecOnApp positions i brecOnConst FArgs r v
      -- Abstract over the fixed prefixed
      let valuesNew ← valuesNew.mapM (mkLambdaFVars xs ·)
      return (Array.zip preDefs valuesNew).map fun ⟨preDef, valueNew⟩ => { preDef with value := valueNew }

def reportTermArg (preDef : PreDefinition) (recArgPos : Nat) : MetaM Unit := do
  if let some ref := preDef.termination.terminationBy?? then
    let fn ← lambdaTelescope preDef.value fun xs _ => mkLambdaFVars xs xs[recArgPos]!
    let termArg : TerminationArgument:= {ref := .missing, structural := true, fn}
    let arity ← lambdaTelescope preDef.value fun xs _ => pure xs.size
    let stx ← termArg.delab arity (extraParams := preDef.termination.extraParams)
    Tactic.TryThis.addSuggestion ref stx

private def inferRecArgPos (preDefs : Array PreDefinition)
    (termArgs? : Option TerminationArguments) : M (Array Nat × Array PreDefinition) := do
  withoutModifyingEnv do
    if let some termArgs := termArgs? then
      let recArgPoss ← termArgs.mapM (·.structuralArg)
      let preDefsNew ← elimMutualRecursion preDefs recArgPoss
      return (recArgPoss, preDefsNew)
    else
      let #[preDef] := preDefs
        | throwError "mutual structural recursion requires explicit `termination_by` clauses"
      -- Use termination_by annotation to find argument to recurse on, or just try all
      tryAllArgs preDef.value fun i =>
        mapError (f := fun msg => m!"argument #{i+1} cannot be used for structural recursion{indentD msg}") do
          let preDefsNew ← elimMutualRecursion #[preDef] #[i]
          return (#[i], preDefsNew)

def structuralRecursion (preDefs : Array PreDefinition) (termArgs? : Option TerminationArguments) : TermElabM Unit := do
  let ((recArgPoss, preDefsNonRec), state) ← run <| inferRecArgPos preDefs termArgs?
  for recArgPos in recArgPoss, preDef in preDefs do
    reportTermArg preDef recArgPos
  state.addMatchers.forM liftM
  preDefsNonRec.forM fun preDefNonRec => do
    let preDefNonRec ← eraseRecAppSyntax preDefNonRec
    -- state.addMatchers.forM liftM
    mapError (addNonRec preDefNonRec (applyAttrAfterCompilation := false)) fun msg =>
      m!"structural recursion failed, produced type incorrect term{indentD msg}"
    -- We create the `_unsafe_rec` before we abstract nested proofs.
    -- Reason: the nested proofs may be referring to the _unsafe_rec.
  let preDefs ← preDefs.mapM (eraseRecAppSyntax ·)
  addAndCompilePartialRec preDefs
  for preDef in preDefs, recArgPos in recArgPoss do
    let mut preDef := preDef
    unless preDef.kind.isTheorem do
      unless (← isProp preDef.type) do
        preDef ← abstractNestedProofs preDef
        /-
        Don't save predefinition info for equation generator
        for theorems and definitions that are propositions.
        See issue #2327
        -/
        registerEqnsInfo preDef (preDefs.map (·.declName)) recArgPos
    addSmartUnfoldingDef preDef recArgPos
    markAsRecursive preDef.declName
  applyAttributesOf preDefsNonRec AttributeApplicationTime.afterCompilation

builtin_initialize
  registerTraceClass `Elab.definition.structural

end Structural

export Structural (structuralRecursion)

end Lean.Elab
