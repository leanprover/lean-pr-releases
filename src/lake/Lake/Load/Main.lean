/-
Copyright (c) 2022 Mac Malone. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Mac Malone, Gabriel Ebner
-/
import Lake.Util.EStateT
import Lake.Util.StoreInsts
import Lake.Config.Workspace
import Lake.Build.Topological
import Lake.Build.Module
import Lake.Build.Package
import Lake.Build.Library
import Lake.Load.Materialize
import Lake.Load.Package
import Lake.Load.Elab
import Lake.Load.Toml

open System Lean

/-! # Loading a Workspace
The main definitions for loading a workspace and resolving dependencies.
-/

namespace Lake

/--
Elaborate a dependency's Lean configuration file into a `Package`.
The resulting package does not yet include any dependencies.
-/
def loadLeanConfig (cfg : LoadConfig)
: LogIO (Package × Environment) := do
  let configEnv ← importConfigFile cfg
  let pkgConfig ← IO.ofExcept <| PackageConfig.loadFromEnv configEnv cfg.leanOpts
  let pkg : Package := {
    dir := cfg.pkgDir
    relDir := cfg.relPkgDir
    config := pkgConfig
    relConfigFile := cfg.relConfigFile
    remoteUrl? := cfg.remoteUrl?
  }
  return (← pkg.loadFromEnv configEnv cfg.leanOpts, configEnv)

/--
Return whether a configuration file with the given name
and/or a supported extension exists.
-/
def configFileExists (cfgFile : FilePath) : BaseIO Bool :=
  if cfgFile.extension.isSome then
    cfgFile.pathExists
  else
    let leanFile := cfgFile.addExtension "lean"
    let tomlFile := cfgFile.addExtension "toml"
    leanFile.pathExists <||> tomlFile.pathExists

/-- Loads a Lake package configuration (either Lean or TOML). -/
def loadPackageCore
  (name : String) (cfg : LoadConfig)
: LogIO (Package × Option Environment) := do
  if let some ext := cfg.relConfigFile.extension then
    unless (← cfg.configFile.pathExists) do
      error s!"{name}: configuration file not found: {cfg.configFile}"
    match ext with
    | "lean" => (·.map id some) <$> loadLeanConfig cfg
    | "toml" => ((·,none)) <$> loadTomlConfig cfg.pkgDir cfg.relPkgDir cfg.relConfigFile
    | _ => error s!"{name}: configuration has unsupported file extension: {cfg.configFile}"
  else
    let relLeanFile := cfg.relConfigFile.addExtension "lean"
    let relTomlFile := cfg.relConfigFile.addExtension "toml"
    let leanFile := cfg.pkgDir / relLeanFile
    let tomlFile := cfg.pkgDir / relTomlFile
    let leanExists ← leanFile.pathExists
    let tomlExists ← tomlFile.pathExists
    if leanExists then
      if tomlExists then
        logInfo s!"{name}: {relLeanFile} and {relTomlFile} are both present; using {relLeanFile}"
      (·.map id some) <$> loadLeanConfig {cfg with relConfigFile := relLeanFile}
    else
      if tomlExists then
        ((·,none)) <$> loadTomlConfig cfg.pkgDir cfg.relPkgDir relTomlFile
      else
        error s!"{name}: no configuration file with a supported extension:\n{leanFile}\n{tomlFile}"

/--
Loads the package configuration of a materialized dependency.
Adds the facets defined in the package to the `Workspace`.
-/
def loadDepPackage
  (dep : MaterializedDep)
  (lakeOpts : NameMap String) (leanOpts : Options) (reconfigure : Bool)
: StateT Workspace LogIO Package := fun ws => do
  let name := dep.name.toString (escape := false)
  let (pkg, env?) ← loadPackageCore name {
    lakeEnv := ws.lakeEnv
    wsDir := ws.dir
    relPkgDir := dep.relPkgDir
    relConfigFile := dep.configFile
    lakeOpts, leanOpts, reconfigure
    remoteUrl? := dep.remoteUrl?
  }
  if let some env := env? then
    let ws ← IO.ofExcept <| ws.addFacetsFromEnv env leanOpts
    return (pkg, ws)
  else
    return (pkg, ws)

/--
Load a `Workspace` for a Lake package by elaborating its configuration file.
Does not resolve dependencies.
-/
def loadWorkspaceRoot (config : LoadConfig) : LogIO Workspace := do
  Lean.searchPathRef.set config.lakeEnv.leanSearchPath
  let (root, env?) ← loadPackageCore "[root]" config
  let ws : Workspace := {
    root, lakeEnv := config.lakeEnv
    moduleFacetConfigs := initModuleFacetConfigs
    packageFacetConfigs := initPackageFacetConfigs
    libraryFacetConfigs := initLibraryFacetConfigs
  }
  if let some env := env? then
    IO.ofExcept <| ws.addFacetsFromEnv env config.leanOpts
  else
    return ws

/-- Loads a Lake package as a single independent object (without dependencies). -/
def loadPackage (config : LoadConfig) : LogIO Package := do
  Lean.searchPathRef.set config.lakeEnv.leanSearchPath
  (·.1) <$> loadPackageCore "[root]" config

/-- The monad of the dependency resolver. -/
abbrev ResolveT m := CallStackT Name <| StateT Workspace m

/-- Log dependency cycle and error. -/
@[specialize] def depCycleError [MonadError m] (cycle : Cycle Name) : m α :=
  error s!"dependency cycle detected:\n{"\n".intercalate <| cycle.map (s!"  {·}")}"

instance [Monad m] [MonadError m] : MonadCycleOf Name (ResolveT m) where
  throwCycle := depCycleError

/--
Recursively visits the workspace dependency graph, starting from `root`.
At each package, performs `f` to resolve the dependencies of a package,
recurses, then adds the package to workspace's package set.
Errors if a cycle is encountered.

**Traversal Order**

All dependencies of a package are visited in order before recursing to the
dependencies' dependencies. For example, given the dependency graph:

```
R
|- A
|- B
 |- X
 |- Y
|- C
```

Lake follows the order `R`, `A`, `B`, `C`, `X`, `Y`.

The logic behind this design is that users would expect the dependencies
they write in a package configuration to be resolved accordingly and would be
surprised if they are overridden by nested dependencies referring to the same
package.

For example, were Lake to use a pure depth-first traversal, Lake would follow
the order `R`, `A`, `B`, `X`, `Y`, `C`. If `X` and `C` are both the package
`foo`, Lake would use the configuration of `foo` found in `B` rather than in
the root `R`, which would likely confuse the user.
-/
@[inline] def Workspace.resolveDeps
  [Monad m] [MonadError m] (ws : Workspace)
  (f : Package → Dependency → ResolveT m Package)
: m Workspace := do
  let (root, ws) ←
    StateT.run (s := ws) <|
    ReaderT.run (r := []) <|
    StateT.run' (s := {}) <|
    recFetchAcyclic (·.name) go ws.root
  return {ws with root}
where
  @[specialize] go pkg resolve : StateT (NameMap Package) (ResolveT m) Package := do
    pkg.depConfigs.forM fun dep => do
      if (← getThe (NameMap Package)).contains dep.name then
        return
      if (← getThe Workspace).packageMap.contains dep.name then
        return -- already resolved in another branch
      if pkg.name = dep.name then
        error s!"{pkg.name}: package requires itself (or a package with the same name)"
      let dep ← f pkg dep -- package w/o dependencies
      store dep.name dep
    let deps ← pkg.depConfigs.mapM fun dep => do
      if let some dep ← fetch? dep.name then
        modifyThe (NameMap Package) (·.erase dep.name) -- for `dep` linearity
        return OpaquePackage.mk (← resolve dep)
      if let some dep ← findPackage? dep.name then
        return OpaquePackage.mk dep -- already resolved in another branch
      error s!"{dep.name}: impossible resolution state reached"
    let pkg := {pkg with opaqueDeps := deps}
    modifyThe Workspace (·.addPackage pkg)
    return pkg

def stdMismatchError (newName : String) (rev : String) :=
s!"the 'std' package has been renamed to '{newName}' and moved to the
'leanprover-community' organization; downstream packages which wish to
update to the new std should replace

  require std from
    git \"https://github.com/leanprover/std4\"{rev}

in their Lake configuration file with

  require {newName} from
    git \"https://github.com/leanprover-community/{newName}\"{rev}

"

/--
Rebuild the workspace's Lake manifest and materialize missing dependencies.

Packages are updated to latest specific revision matching that in `require`
(e.g., if the `require` is `@master`, update to latest commit on master) or
removed if the `require` is removed. If `tuUpdate` is empty, update/remove all
root dependencies. Otherwise, only update the root dependencies specified.

If `reconfigure`, elaborate configuration files while updating, do not use OLeans.
-/
def Workspace.updateAndMaterialize
  (ws : Workspace) (leanOpts : Options := {})
  (toUpdate : NameSet := {}) (reconfigure := true)
: LogIO Workspace := do
  let (ws, entries) ← StateT.run (s := mkNameMap PackageEntry) do
    -- Use manifest versions of root packages that should not be updated
    let rootName := ws.root.name.toString (escape := false)
    match (← Manifest.load ws.manifestFile |>.toBaseIO) with
    | .ok manifest =>
      unless toUpdate.isEmpty do
        manifest.packages.forM fun entry => do
          unless entry.inherited || toUpdate.contains entry.name do
            modifyThe (NameMap PackageEntry) (·.insert entry.name entry)
      if let some oldRelPkgsDir := manifest.packagesDir? then
        let oldPkgsDir := ws.dir / oldRelPkgsDir
        if oldRelPkgsDir.normalize != ws.relPkgsDir.normalize && (← oldPkgsDir.pathExists) then
          logInfo s!"workspace packages directory changed; \
            renaming '{oldPkgsDir}' to '{ws.pkgsDir}'"
          let doRename : IO Unit := do
            createParentDirs ws.pkgsDir
            IO.FS.rename oldPkgsDir ws.pkgsDir
          if let .error e ← doRename.toBaseIO then
            error s!"could not rename workspace packages directory: {e}"
    | .error (.noFileOrDirectory ..) =>
      logInfo s!"{rootName}: no previous manifest, creating one from scratch"
    | .error e =>
      unless toUpdate.isEmpty do
        liftM (m := IO) <| throw e -- only ignore manifest on a bare `lake update`
      logWarning s!"{rootName}: ignoring previous manifest because it failed to load: {e}"
    ws.resolveDeps fun pkg dep => do
      let ws ← getThe Workspace
      let inherited := pkg.name != ws.root.name
      -- Materialize the dependency
      let matDep ← id do
        if let some entry ← fetch? dep.name then
          entry.materialize ws.lakeEnv ws.dir ws.relPkgsDir
        else
          let matDep ← dep.materialize inherited ws.lakeEnv ws.dir ws.relPkgsDir pkg.relDir
          store matDep.name matDep.manifestEntry
          return matDep
      -- Load the package
      let depPkg ← loadDepPackage matDep dep.opts leanOpts reconfigure
      if depPkg.name ≠ dep.name then
        if dep.name = .mkSimple "std" then
          let rev :=
            match matDep.manifestEntry.src with
            | .git (inputRev? := some rev) .. => s!" @ {repr rev}"
            | _ => ""
          logError (stdMismatchError depPkg.name.toString rev)
        if let .error e ← IO.FS.removeDirAll depPkg.dir |>.toBaseIO then -- cleanup
          -- Deleting git repositories via IO.FS.removeDirAll does not work reliably on Windows
          logError s!"'{dep.name}' was downloaded incorrectly; \
            you will need to manually delete '{depPkg.dir}': {e}"
        error s!"{pkg.name}: package '{depPkg.name}' was required as '{dep.name}'"
      -- Add the dependencies' locked dependencies to the manifest
      match (← Manifest.load depPkg.manifestFile |>.toBaseIO) with
      | .ok manifest =>
        manifest.packages.forM fun entry => do
          unless (← getThe (NameMap PackageEntry)).contains entry.name do
            let entry := entry.setInherited.inDirectory depPkg.relDir
            store entry.name entry
      | .error (.noFileOrDirectory ..) =>
        logWarning s!"{depPkg.name}: ignoring missing dependency manifest '{depPkg.manifestFile}'"
      | .error e =>
        logWarning s!"{depPkg.name}: ignoring dependency manifest because it failed to load: {e}"
      return depPkg
  let manifestEntries := ws.packages.foldl (init := #[]) fun arr pkg =>
    match entries.find? pkg.name with
    | some entry => arr.push <|
      entry.setManifestFile pkg.relManifestFile |>.setConfigFile pkg.relConfigFile
    | none => arr -- should only be the case for the root
  let manifest : Manifest := {
    name := ws.root.name
    lakeDir := ws.relLakeDir
    packagesDir? := ws.relPkgsDir
    packages := manifestEntries
  }
  manifest.saveToFile ws.manifestFile
  LakeT.run ⟨ws⟩ <| ws.packages.forM fun pkg => do
    unless pkg.postUpdateHooks.isEmpty do
      logInfo s!"{pkg.name}: running post-update hooks"
      pkg.postUpdateHooks.forM fun hook => hook.get.fn pkg
  return ws

/--
Check whether the manifest exists and
whether entries in the manifest are up-to-date,
reporting warnings and/or errors as appropriate.
-/
def validateManifest
  (pkgEntries : NameMap PackageEntry) (deps : Array Dependency)
: LogIO PUnit := do
  if pkgEntries.isEmpty && !deps.isEmpty then
    error "missing manifest; use `lake update` to generate one"
  deps.forM fun dep => do
    let warnOutOfDate (what : String) :=
      logWarning <|
        s!"manifest out of date: {what} of dependency '{dep.name}' changed; \
        use `lake update {dep.name}` to update it"
    if let .some entry := pkgEntries.find? dep.name then
    match dep.src, entry.src with
    | .git (url := url) (rev := rev) .., .git (url := url') (inputRev? := rev')  .. =>
      if url ≠ url' then warnOutOfDate "git url"
      if rev ≠ rev' then warnOutOfDate "git revision"
    | .path .., .path .. => pure ()
    | _, _ => warnOutOfDate "source kind (git/path)"

/--
Resolving a workspace's dependencies using a manifest,
downloading and/or updating them as necessary.
-/
def Workspace.materializeDeps
  (ws : Workspace) (manifest : Manifest)
  (leanOpts : Options := {}) (reconfigure := false)
: LogIO Workspace := do
  if !manifest.packages.isEmpty && manifest.packagesDir? != some (mkRelPathString ws.relPkgsDir) then
    logWarning <|
      "manifest out of date: packages directory changed; \
      use `lake update` to rebuild the manifest \
      (warning: this will update ALL workspace dependencies)"
  let relPkgsDir := manifest.packagesDir?.getD ws.relPkgsDir
  let pkgEntries : NameMap PackageEntry := manifest.packages.foldl (init := {})
    fun map entry => map.insert entry.name entry
  validateManifest pkgEntries ws.root.depConfigs
  ws.resolveDeps fun pkg dep => do
    let ws ← getThe Workspace
    if let some entry := pkgEntries.find? dep.name then
      let result ← entry.materialize ws.lakeEnv ws.dir relPkgsDir
      loadDepPackage result dep.opts leanOpts reconfigure
    else
      if pkg.name = ws.root.name then
        error <|
          s!"dependency '{dep.name}' not in manifest; \
          use `lake update {dep.name}` to add it"
      else
        error <|
          s!"dependency '{dep.name}' of '{pkg.name}' not in manifest; \
          this suggests that the manifest is corrupt; \
          use `lake update` to generate a new, complete file \
          (warning: this will update ALL workspace dependencies)"

/--
Load a `Workspace` for a Lake package by
elaborating its configuration file and resolving its dependencies.
If `updateDeps` is true, updates the manifest before resolving dependencies.
-/
def loadWorkspace (config : LoadConfig) (updateDeps := false) : LogIO Workspace := do
  let rc := config.reconfigure
  let leanOpts := config.leanOpts
  let ws ← loadWorkspaceRoot config
  if updateDeps then
    ws.updateAndMaterialize leanOpts {} rc
  else if let some manifest ← Manifest.load? ws.manifestFile then
    ws.materializeDeps manifest leanOpts rc
  else
    ws.updateAndMaterialize leanOpts {} rc

/-- Updates the manifest for the loaded Lake workspace (see `updateAndMaterialize`). -/
def updateManifest (config : LoadConfig) (toUpdate : NameSet := {}) : LogIO Unit := do
  let rc := config.reconfigure
  let leanOpts := config.leanOpts
  let ws ← loadWorkspaceRoot config
  discard <| ws.updateAndMaterialize leanOpts toUpdate rc
