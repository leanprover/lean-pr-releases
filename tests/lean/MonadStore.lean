import Lake.Load.Main

open Lake

-- when `MonadStore1` had an outParam, it was not possible for the following two instances to synthesize

#synth MonadStore1 `key Int $ CycleT Name
  (StateT (NameMap Nat) (StateT (NameMap Int) IO))

#synth MonadStore1 `key Nat $ CycleT Name
  (StateT (NameMap Nat) (StateT (NameMap Int) IO))
