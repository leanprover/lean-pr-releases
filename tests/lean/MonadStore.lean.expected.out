instMonadStore1OfMonadStore
instMonadStore1OfMonadStore
